### PR TITLE
chapter 13

### PR DESCRIPTION
**Pull request by chan227**

### CODE REVIEW

**Comment by chan227** [comment]
ready

File: app/assets/stylesheets/microposts.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Microposts controller here.
+// They will automatically be included in application.css.
```
Comment (saobv-2223) [inline app/assets/stylesheets/microposts.scss]: remove file that never used

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -18,4 +18,14 @@ def extract_locale
...
   end
+
+  private
+
+  def logged_in_user
+    return if logged_in?
+
+      store_location
+      flash[:danger] = t ".login_remind"
+      redirect_to login_path
```
Comment (saobv-2223) [inline app/controllers/application_controller.rb]:
```suggestion
    store_location
    flash[:danger] = t ".login_remind"
    redirect_to login_path
```

File: app/helpers/microposts_helper.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module MicropostsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/microposts_helper.rb]: remove file that never used

File: app/models/micropost.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class Micropost < ApplicationRecord
+  belongs_to :user
+  scope :recent_posts, ->{order created_at: :desc}
+  has_one_attached :image
+  validates :user_id, presence: true
```
Comment (saobv-2223) [inline app/models/micropost.rb]: image

File: app/views/shared/_micropost_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+<% end %>
+
+<script type="text/javascript">
+  $("#micropost_image").bind("change", function() {
+    var size_in_megabytes = this.files[0].size/1024/1024;
+    if (size_in_megabytes > 5) {
+      alert("Maximum file size is 5MB. Please choose a smaller file.");
+    }
+  });
+</script>
```
Comment (saobv-2223) [inline app/views/shared/_micropost_form.html.erb]: Don't write js in html file. Therefore, you should write in another .js file in pack folder and required to application.js

Comment (chan227) [reply app/views/shared/_micropost_form.html.erb]: done

File: app/models/micropost.rb (hunk 4 above)
Comment (chan227) [reply app/models/micropost.rb]: done

File: app/helpers/microposts_helper.rb (hunk 3 above)
Comment (chan227) [reply app/helpers/microposts_helper.rb]: done

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (chan227) [reply app/controllers/application_controller.rb]: done

File: app/assets/stylesheets/microposts.scss (hunk 1 above)
Comment (chan227) [reply app/assets/stylesheets/microposts.scss]: done

**Comment by chan227** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

File: app/controllers/microposts_controller.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    else
+      flash.now[:danger] = t ".fail_mess"
+      @feed_items = current_user.feed.paginate page: params[:page]
+      render "static_pages/home"
+    end
+  end
+
+  def destroy
+    @micropost.destroy
+    flash[:success] = t ".success_mess"
```
Comment (vanvtt-0952) [inline app/controllers/microposts_controller.rb]: plz, add code to handle "destroy fail"

File: app/controllers/microposts_controller.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+  private
+
+  def micropost_params
+    params.require(:micropost).permit :content, :image
+  end
+
+  def correct_user
+    @micropost = current_user.microposts.find_by id: params[:id]
+    redirect_to root_path unless @micropost
```
Comment (vanvtt-0952) [inline app/controllers/microposts_controller.rb]: add a message when going "redirect root"

File: app/controllers/microposts_controller.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+  def create
+    @micropost = current_user.microposts.build micropost_params
+    @micropost.image.attach micropost_params[:image]
+    if @micropost.save
+      flash[:success] = t ".success_mess"
+      redirect_to root_path
+    else
+      flash.now[:danger] = t ".fail_mess"
+      @feed_items = current_user.feed.paginate page: params[:page]
```
Comment (vanvtt-0952) [inline app/controllers/microposts_controller.rb]: change 
` @feed_items = current_user.feed.paginate page: params[:page]`
to 
` @feed_items = current_user.feed.recent_posts.paginate page: params[:page]`

**Pull request merged**